prow/test/integration/fakeghserver/fakeghserver.go

```diff
@@ -0,0 +1,258 @@
+	r.Path("/repos/{org}/{repo}/commits/{sha}/status").Queries("per_page", "{page}").Handler(statusHandler(ghClient))
+	r.Path("/repos/{org}/{repo}/issues").Handler(issueHandler(ghClient))
+	r.Path("/repos/{org}/{repo}/issues/{issue_id}/comments").Handler(issueCommentHandler(ghClient))
+	r.Path("/repos/{org}/{repo}/issues/comments/${comment_id}").Handler(issueCommentHandler(ghClient))
+
+	health := pjutil.NewHealth()
+	health.ServeReady()
+
+	logrus.Info("Start server")
+
+	// setup done, actually start the server
+	server := &http.Server{Addr: ":8888", Handler: r}
+	interrupts.ListenAndServe(server, 5*time.Second)
+}
+
+func unmarshal(w http.ResponseWriter, r *http.Request, data interface{}) error {
+	d := json.NewDecoder(r.Body)
+	d.DisallowUnknownFields()
+
+	w.Header().Set("Content-Type", "application/json")
+	err := d.Decode(&data)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "{\"error\": \"Failed unmarshal request: %v\"}", err.Error())
+	}
+	return err
+}
+
+func response(w http.ResponseWriter, r *http.Request, msg string, ok int, err error) error {
+	logrus.Infof("request: %s - %s. responses: %s, %d, %v", r.URL.Path, r.Method, msg, ok, err)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprint(w, err.Error())
+		logrus.Info(err)
+		return err
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.Header().Set("Link", "")
+	w.WriteHeader(ok)
+	fmt.Fprint(w, msg)
+	logrus.Info("Succeeded with request: ", ok)
+	return nil
+}
+
+func defaultHandler() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		logrus.Infof("Not supported: %s, %s", r.URL.Path, r.Method)
+		if err := response(w, r, "", http.StatusNotFound,
+			fmt.Errorf("{\"error\": \"API not supported\"}, %s, %s", r.URL.Path, r.Method)); err != nil {
+			logrus.WithError(err).Error("failed serving default handler")
+		}
+	})
+}
+
+func userHandler(ghc *fakegithub.FakeClient) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		logrus.Infof("Serving: %s, %s", r.URL.Path, r.Method)
+		var msg string
+		var ok int
+		var err error
+		ok = http.StatusOK
+		userData, err := ghc.BotUser()
+		if err == nil {
+			var content []byte
+			content, err = json.Marshal(&userData)
+			msg = string(content)
+		}
+		if err := response(w, r, msg, ok, err); err != nil {
+			logrus.WithError(err).Error("failed serving user handler")
+		}
+	})
+}
+
+func statusHandler(ghc *fakegithub.FakeClient) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		logrus.Infof("Serving: %s, %s", r.URL.Path, r.Method)
+		var msg string
+		var ok int
+		var err error
+		ok = http.StatusOK
+		vars := mux.Vars(r)
+		org, repo, SHA := vars["org"], vars["repo"], vars["sha"]
+		if r.Method == http.MethodPost {
+			ok = http.StatusCreated
+			data := prowgh.Status{}
+			err = unmarshal(w, r, &data)
+			if err == nil {
```

Comment:
In order to avoid the weird `if err == nil { err = doNextThing() }` nesting, it might be worth to implement `response` as a wrapper that takes the actual handler as a `func(r *http.Request) (statusCode int, err error)`

Reply:
Good idea, implemented, and looks a lot nicer